### PR TITLE
KAFKA-14285: Delete quota node in zookeeper when configs are empty

### PR DESCRIPTION
**Pull request by aglicacha**
The PR resolve issue KAFKA-14285. After doing changeConfigs, check and clean quota nodes if configs are empty, to avoid infinite increasement of quota nodes in zookeeper
related PR: https://github.com/apache/kafka/pull/12727
@showuon

### CODE REVIEW

**Comment by aglicacha** [comment]
@soarez Ready to review.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (soarez) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala]: Should we add a line or two with a comment explaining why we're adding this change here?

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (soarez) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala]:
```suggestion
          if (clientsChildren == Seq(clientId)) {
```

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (soarez) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala]:
```suggestion
            if (userData.isEmpty && userChildren == Seq(ConfigType.Client)) {
```

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (soarez) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala]: Looking at this block of code, there are apparently three different scenarios for the value of `pathToDelete`, from lines:

1. 372 `var pathToDelete = currPath`
2. 379 `pathToDelete = clientsPath`
3. 384 `pathToDelete = userPath`

 But following through the test cases we only seem to hit two of those. `testChangeConfigsWithUserAndClientId` hits 3. and all the other current tests stick to 1. Should we have an extra test case to cover scenario 2. or, if that scenario should never happen, should this logic be re-structured?

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (soarez) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala:370]: It seems that the switch/match block above ends up calling `changeEntityConfig` regardless of the entity type. As a final step, `changeEntityConfig` creates a ZK notification. Would it be better to perform the cleanup before the ZK notification is triggered?

**Review by soarez — COMMENTED** [review]
Thanks for the contribution @aglicacha. I've left some comments.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala:370]: Agree with that. I have moved the clean logic to method changeEntityConfig before zk notification is triggered.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala]: Yeah, I have added some annotation to explain it.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala]: Done.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala]: Done.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala]: Right. Scenario 2 will happen if the user configs are not empty but clients under the user is empty. I have added test case to cover this scenario. Thank you.

**Comment by aglicacha** [comment]
@soarez Ready to review.

**Review by soarez — APPROVED** [review]
LGTM

@mimaison could you review this?

**Comment by aglicacha** [comment]
@mimaison @showuon  could you review this, please?

**Comment by mimaison** [comment]
Sorry for the delay, I'll try to take a look this week.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (showuon) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala:359]: Please update the javddoc above.

File: core/src/test/scala/unit/kafka/zk/AdminZkClientTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/zk/AdminZkClientTest.scala]: The 1st parameter of assertEquals method should be `expected result`. Same comments also apply to below.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (showuon) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala:404]: Should we log here about zk node deleted?

**Review by showuon — COMMENTED** [review]
@aglicacha , thanks for the PR. Had a first pass, left some comments. Thanks.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (showuon) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala]: nit: which means [if] the configs are ....

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (showuon) [inline core/src/main/scala/kafka/zk/AdminZkClient.scala]: Could we add some javadoc above this method? And also some inline comments for `isUserClientId` case below, to explain what nodes we're trying to delete for each case. Thanks.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala:359]: Done.

File: core/src/test/scala/unit/kafka/zk/AdminZkClientTest.scala
Comment (aglicacha) [reply core/src/test/scala/unit/kafka/zk/AdminZkClientTest.scala]: Done.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala:404]: Done.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala]: Done.

File: core/src/main/scala/kafka/zk/AdminZkClient.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/zk/AdminZkClient.scala]: Done.

**Comment by aglicacha** [comment]
@showuon Thanks for the comments. I have fixed them. Please to review again, thanks.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala
Comment (dengziming) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: I think we can use `isUserClientId = entity.child.exists(e => ConfigType.Client.equals(e.entityType))` directly?

**Review by dengziming — APPROVED** [review]
Thanks for this fix @aglicacha , can you also test this change locally and get the ZK data before and after this change, and add them in the PR description to make sure there is no harmful effect.

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the patch! I'll merge it after you addressing @dengziming 's comments.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala
Comment (aglicacha) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: Right, thanks for the advice. I have fixed it.

**Comment by aglicacha** [comment]
Add some test results with zk data before and after this change: 

test1:
	operations:
		1.add quota(producer_byte_rate=1024) for user01, user02, user03
		2.add quota(producer_byte_rate=1024) for client01, client02, client03
		3.add quota(producer_byte_rate=1024) for userClient user01/client11, user01/client12, user01/client13
		4.delete quota(producer_byte_rate) for user01, user02, user03
		5.delete quota(producer_byte_rate) for client01, client02, client03
		6.delete quota(producer_byte_rate) for userClient user01/client11, user01/client12, user01/client13

	zk data result(before this change):
		/config
		 |-- /users: null
			  |-- /user01: {"version":1,"config":{}}
				   |-- /clients: null
					    |-- /client11: {"version":1,"config":{}}
					    |-- /client12: {"version":1,"config":{}}
					    |-- /client13: {"version":1,"config":{}}
			  |-- /user02: {"version":1,"config":{}}
			  |-- /user03: {"version":1,"config":{}}
		 |-- /clients: null
		      |-- client01: {"version":1,"config":{}}
	              |-- client02: {"version":1,"config":{}}
		      |-- client03: {"version":1,"config":{}}
			
	zk data result(after this change):
		/config
		 |-- /users: null
		 |-- /clients: null
		 
test2:
	operations:
		1.add quota(producer_byte_rate=1024) for user01, user02
		2.add quota(producer_byte_rate=1024) for userClient user01/client01, user02/client02
		3.delete quota(producer_byte_rate) for userClient user01/client01
		4.delete quota(producer_byte_rate) for user02
		
	zk data result(before this change):
		/config
		 |-- /users: null
			  |-- /user01: {"version":1,"config":{"producer_byte_rate":"1024"}}
				   |-- /clients: null
					    |-- /client01: {"version":1,"config":{}}
			  |-- /user02: {"version":1,"config":{}}
				   |-- /clients: null
					    |-- /client02: {"version":1,"config":{"producer_byte_rate":"1024"}}
		 |-- /clients: null
	
	zk data result(after this change):
		/config
		 |-- /users: null
			  |-- /user01: {"version":1,"config":{"producer_byte_rate":"1024"}}
			  |-- /user02: {"version":1,"config":{}}
				   |-- /clients: null
					    |-- /client02: {"version":1,"config":{"producer_byte_rate":"1024"}}
		 |-- /clients: null
		 
test3:
	operations:
		1.add quota(producer_byte_rate=1024) for userClient user01/client01, user01/client02	
		2.delete quota(producer_byte_rate) for userClient user01/client01

	zk data result(before this change):
		/config
		 |-- /users: null
			  |-- /user01: null
				   |-- /clients: null
					    |-- /client01: {"version":1,"config":{}}
					    |-- /client02: {"version":1,"config":{"producer_byte_rate":"1024"}}
		 |-- /clients: null
		 
	zk data result(after this change):
		/config
		 |-- /users: null
			  |-- /user01: null
				   |-- /clients: null
					    |-- /client02: {"version":1,"config":{"producer_byte_rate":"1024"}}
		 |-- /clients: null

All test results are in line with expectations.

**Comment by aglicacha** [comment]
> Thanks for this fix @aglicacha , can you also test this change locally and get the ZK data before and after this change, and add them in the PR description to make sure there is no harmful effect.

I have added some test results in the comment.
@dengziming

**Review by dengziming — APPROVED** [review]
LTGM, thank you for your patience @aglicacha .

**Comment by showuon** [comment]
@aglicacha , there is compile error, could you take a look?

**Comment by aglicacha** [comment]
> @aglicacha , there is compile error, could you take a look?

The error was caused by the unnecessary default value of param isUserClientId in method tryCleanQuotaNodes. I have fixed it. Thanks.
@showuon

**Comment by showuon** [comment]
Thanks for the update. It failed many tests in jdk17, should be infra's issue. Rerunning the tests: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12729/7/

**Comment by aglicacha** [comment]
> Thanks for the update. It failed many tests in jdk17, should be infra's issue. Rerunning the tests: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12729/7/

The test failures do not seem to be caused by the change of this PR. I'm not familiar with it. Any advice for handling this? Thanks.

**Comment by showuon** [comment]
Failed tests are unrelated
```
    Build / JDK 11 and Scala 2.13 / kafka.api.PlaintextEndToEndAuthorizationTest.testNoConsumeWithoutDescribeAclViaSubscribe(String).quorum=zk
    Build / JDK 8 and Scala 2.12 / org.apache.kafka.streams.integration.EosIntegrationTest.shouldWriteLatestOffsetsToCheckpointOnShutdown[exactly_once]
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.SmokeTestDriverIntegrationTest.[1] true
```

**Comment by showuon** [comment]
Thanks for the contribution, @aglicacha !

**Comment by vongosling** [comment]
Good catch, after merging the trunks, I wonder which versions have been resolved, and I see that a large number of users are using 1.1.0 :-)

**Comment by showuon** [comment]
This issue will be fixed in v3.4.1, v3.5.0 and later. Thanks.

**Pull request merged**
